File: Projects/Football/PredictWinner/random_forest_model.R
options(scipen=20, digits=4)

# Load packages
library(randomForest)  # version 4.6-12

#======================================================================================================
# Load Data (Assumes your current working directory is the Football project directory)

train <- read.csv("Data/train.csv")
test <- read.csv("Data/test.csv")

#======================================================================================================
# Format the training data to the specifications for randomForest
# See help(randomForest)

#--------------------------------------------------
# Convert boolean features to factors (i.e. categorical features with levels {TRUE, FALSE}. You could also
# convert to intgers {0, 1})

train$SaintsAtHome <- factor(train$SaintsAtHome, levels=c("FALSE", "TRUE"))
train$Expert1PredWin <- factor(train$Expert1PredWin, levels=c("FALSE", "TRUE"))
train$Expert2PredWin <- factor(train$Expert2PredWin, levels=c("FALSE", "TRUE"))
train$SaintsWon <- factor(train$SaintsWon, levels=c("FALSE", "TRUE"))
test$SaintsAtHome <- factor(test$SaintsAtHome, levels=c("FALSE", "TRUE"))
test$Expert1PredWin <- factor(test$Expert1PredWin, levels=c("FALSE", "TRUE"))
test$Expert2PredWin <- factor(test$Expert2PredWin, levels=c("FALSE", "TRUE"))

#--------------------------------------------------
# Split the training features from the target

train_X <- train[, c("Opponent", "OppRk", "SaintsAtHome", "Expert1PredWin", "Expert2PredWin")]
test_X <- test[, c("Opponent", "OppRk", "SaintsAtHome", "Expert1PredWin", "Expert2PredWin")]
test_X$Opponent <- factor(test_X$Opponent, levels=levels(train_X$Opponent))  # Make sure the levels of Opponent in test are the same as train
train_y <- train$SaintsWon

#======================================================================================================
# Build some random forest models

#--------------------------------------------------
# Random Forest with 101 trees and the rest defaults

# Train the model
rf <- randomForest(x=train_X, y=train_y, ntree=101)

# View results. Confusion matrix based on out of bag samples applied to each tree
rf

# View the 2nd tree
getTree(rf, k=2, labelVar=TRUE)  # see help(getTree) for details

# Make predictions on the test set
predict(rf, test_X)  # class predictions
predict(rf, test_X, type="prob")  # probabilities

# What would the model predict for each training sample
data.frame(Truth=train$SaintsWon, Fitted=predict(rf, train_X), FittedProbTRUE=predict(rf, train_X, type="prob")[, 2])

#--------------------------------------------------
# Random Forest with 101 trees, testing 3 features at each split and keeping track of variable importance

# Train the model
rf <- randomForest(x=train_X, y=train_y, ntree=101, mtry=3, importance=TRUE)

# View the importance of each feature. See help(importance.randomForest)
importance(rf, type=1)  # type=1 -> mean decrease in accuracy
importance(rf, type=2)  # type=2 -> mean decrease in node impurity

#--------------------------------------------------
# Browse more properties of the model

str(rf)

File: Problems/Predict NFL Game Winner/random_forest_model.py
# Imports
import pandas as pd
from sklearn.ensemble import RandomForestClassifier

#======================================================================================================
# Load Data (Assumes your current working directory is the Predict NFL Game Winner problem directory)

train = pd.read_csv("_Data/train.csv")
test = pd.read_csv("_Data/test.csv")

#======================================================================================================
# Format the training data to the specifications for RandomForestClassifier

#--------------------------------------------------
# Convert categorical features to pandas Category type

train.Opponent = pd.Categorical(train.Opponent)
test.Opponent = pd.Categorical(test.Opponent, categories=pd.unique(pd.concat([train.Opponent, test.Opponent])))

#--------------------------------------------------
# Split the training features from the target. Use pd.get_dummies() to one-hot encode categorical values

train_X = train[["OppRk", "SaintsAtHome", "Expert1PredWin", "Expert2PredWin"]]
train_X = pd.concat([train_X, pd.get_dummies(train.Opponent)], axis=1)

test_X = test[["OppRk", "SaintsAtHome", "Expert1PredWin", "Expert2PredWin"]]
test_X = pd.concat([test_X, pd.get_dummies(test.Opponent)], axis=1)

train_y = train.SaintsWon

#======================================================================================================
# Build some random forest models

#--------------------------------------------------
# Random Forest with 101 trees and the rest defaults

# Train the model
rf = RandomForestClassifier(n_estimators=101)
rf.fit(X=train_X, y=train_y)

# View the 2nd tree  (To plot it, see http://scikit-learn.org/stable/modules/tree.html#tree)
rf.estimators_[1]
rf.estimators_[1].feature_importances_  # feature importances for train_X.columns

# Make predictions on the test set
rf.predict(test_X)  # class predictions
rf.predict_proba(test_X)  # probabilities for rf.classes_

# What would the model predict for each training sample
pd.DataFrame({'Truth': train.SaintsWon, 'Fitted': rf.predict(train_X), 'FittedProbTRUE': rf.predict_proba(train_X)[:, 1]})

#--------------------------------------------------
# Random Forest with 101 trees, testing 3 features at each split

# Train the model
rf = RandomForestClassifier(n_estimators=101, max_features=3)
rf.fit(X=train_X, y=train_y)

# View the importance of each feature. See 
rf.feature_importances_  # importances of train_X.columns
